File: akka-skynet/src/main/java/skynet/actors/Profiler.java
package skynet.actors;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.io.Serializable;
import java.util.*;

public class Profiler extends AbstractActor {

	////////////////////////
	// Actor Construction //
	////////////////////////
	
	public static final String DEFAULT_NAME = "profiler";

	public static Props props() {
		return Props.create(Profiler.class);
	}

	////////////////////
	// Actor Messages //
	////////////////////
	
	@Data @AllArgsConstructor
	public static class RegistrationMessage implements Serializable {
		private static final long serialVersionUID = 4545299661052078209L;
	}

	@Data @AllArgsConstructor @SuppressWarnings("unused")
	public static class TaskMessage implements Serializable {
		private static final long serialVersionUID = -8330958742629706627L;
		private TaskMessage() {}
		private int attributes;
	}
	
	@Data @AllArgsConstructor @SuppressWarnings("unused")
	public static class CompletionMessage implements Serializable {
		private static final long serialVersionUID = -6823011111281387872L;
		public enum status {MINIMAL, EXTENDABLE, FALSE, FAILED}
		private CompletionMessage() {}
		private status result;
	}
	
	/////////////////
	// Actor State //
	/////////////////
	
	private final LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    private final Queue<Worker.WorkMessage> unassignedWork = new LinkedList<>();
	private final Queue<ActorRef> idleWorkers = new LinkedList<>();
    private final Map<ActorRef, Worker.WorkMessage> busyWorkers = new HashMap<>();

	private TaskMessage task;

	////////////////////
	// Actor Behavior //
	////////////////////
	
	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(RegistrationMessage.class, this::handle)
				.match(Terminated.class, this::handle)
				.match(TaskMessage.class, this::handle)
				.match(CompletionMessage.class, this::handle)
				.matchAny(object -> this.log.info("Received unknown message: \"{}\"", object.toString()))
				.build();
	}

	private void handle(RegistrationMessage message) {
		this.context().watch(this.sender());
		
		this.assign(this.sender());
		this.log.info("Registered {}", this.sender());
	}
	
	private void handle(Terminated message) {
		this.context().unwatch(message.getActor());
		
		if (!this.idleWorkers.remove(message.getActor())) {
            Worker.WorkMessage work = this.busyWorkers.remove(message.getActor());
			if (work != null) {
				this.assign(work);
			}
		}		
		this.log.info("Unregistered {}", message.getActor());
	}
	
	private void handle(TaskMessage message) {
		if (this.task != null)
			this.log.error("The profiler actor can process only one task in its current implementation!");
		
		this.task = message;
        this.assign(new Worker.WorkMessage(new int[0], new int[0]));
	}
	
	private void handle(CompletionMessage message) {
		ActorRef worker = this.sender();
        Worker.WorkMessage work = this.busyWorkers.remove(worker);

		this.log.info("Completed: [{},{}]", Arrays.toString(work.getX()), Arrays.toString(work.getY()));
		
		switch (message.getResult()) {
			case MINIMAL: 
				this.report(work);
				break;
			case EXTENDABLE:
				this.split(work);
				break;
			case FALSE:
				// Ignore
				break;
			case FAILED:
				this.assign(work);
				break;
		}
		
		this.assign(worker);
	}

    private void assign(Worker.WorkMessage work) {
		ActorRef worker = this.idleWorkers.poll();
		
		if (worker == null) {
			this.unassignedWork.add(work);
			return;
		}
		
		this.busyWorkers.put(worker, work);
		worker.tell(work, this.self());
	}
	
	private void assign(ActorRef worker) {
        Worker.WorkMessage work = this.unassignedWork.poll();
		
		if (work == null) {
			this.idleWorkers.add(worker);
			return;
		}
		
		this.busyWorkers.put(worker, work);
		worker.tell(work, this.self());
	}

    private void report(Worker.WorkMessage work) {
		this.log.info("UCC: {}", Arrays.toString(work.getX()));
	}

    private void split(Worker.WorkMessage work) {
		int[] x = work.getX();
		int[] y = work.getY();
		
		int next = x.length + y.length;
		
		if (next < this.task.getAttributes() - 1) {
			int[] xNew = Arrays.copyOf(x, x.length + 1);
			xNew[x.length] = next;
            this.assign(new Worker.WorkMessage(xNew, y));
			
			int[] yNew = Arrays.copyOf(y, y.length + 1);
			yNew[y.length] = next;
            this.assign(new Worker.WorkMessage(x, yNew));
		}
	}
}

File: src/main/scala/skynet/cluster/actors/Profiler.scala
package skynet.cluster.actors

import java.util

import akka.actor.{AbstractActor, ActorRef, Props, Terminated}
import akka.event.Logging
import skynet.cluster.actors.Profiler.TaskMessage

object Profiler { ////////////////////////
  // Actor Construction //
  ////////////////////////
  val DEFAULT_NAME = "profiler"

  def props: Props = Props.create(classOf[Profiler])

  ////////////////////
  // Actor Messages //
  ////////////////////
  @SerialVersionUID(4545299661052078209L)
  case class RegistrationMessage() {}

  @SerialVersionUID(-8330958742629706627L)
  case class TaskMessage(attributes: Int = 0) {

  }

  @SerialVersionUID(-6823011111281387872L)
  case class CompletionMessage(result: CompletionMessage.CompletionStatus.CompletionStatus) {

  }

  object CompletionMessage {

    object CompletionStatus extends Enumeration {
      type CompletionStatus = Value
      val MINIMAL, EXTENDABLE, FALSE, FAILED = Value
    }

  }

}

class Profiler extends AbstractActor {
  /////////////////
  // Actor State //
  /////////////////
  final private val log = Logging.getLogger(getContext.system, this)
  final private val unassignedWork = new util.LinkedList[Worker.WorkMessage]
  final private val idleWorkers = new util.LinkedList[ActorRef]
  final private val busyWorkers = new util.HashMap[ActorRef, Worker.WorkMessage]
  private var task: TaskMessage = _

  // Actor Behavior //
  override def createReceive: AbstractActor.Receive =
    receiveBuilder
      .`match`(classOf[Profiler.RegistrationMessage], handleRegistration)
      .`match`(classOf[Terminated], handleTermination)
      .`match`(classOf[Profiler.TaskMessage], handleTask)
      .`match`(classOf[Profiler.CompletionMessage], handleCompletion)
      .matchAny((`object`: Any) => this.log.info("Received unknown message: \"{}\"", `object`.toString))
      .build

  private def handleRegistration(message: Profiler.RegistrationMessage): Unit = {
    this.context.watch(sender)
    this.assign(sender)
    this.log.info("Registered {}", sender)
  }

  private def assign(worker: ActorRef): Unit = {
    val work = unassignedWork.poll
    if (work == null) {
      this.idleWorkers.add(worker)
      return
    }
    this.busyWorkers.put(worker, work)
    worker.tell(work, self)
  }

  private def handleTermination(message: Terminated): Unit = {
    this.context.unwatch(message.getActor)
    if (!this.idleWorkers.remove(message.getActor)) {
      val work = this.busyWorkers.remove(message.getActor)
      if (work != null) this.assign(work)
    }
    this.log.info("Unregistered {}", message.getActor)
  }

  private def assign(work: Worker.WorkMessage): Unit = {
    val worker = idleWorkers.poll
    if (worker == null) {
      unassignedWork.add(work)
      return
    }
    this.busyWorkers.put(worker, work)
    worker.tell(work, self)
  }

  private def handleTask(message: Profiler.TaskMessage): Unit = {
    if (this.task != null)
      this.log.error("The profiler actor can process only one task in its current implementation!")

    this.task = message
    this.assign(Worker.WorkMessage(new Array[Int](0), new Array[Int](0)))
  }

  private def handleCompletion(message: Profiler.CompletionMessage): Unit = {
    val worker = this.sender
    val work = this.busyWorkers.remove(worker)
    log.info("Completed: [{},{}]", util.Arrays.toString(work.x), util.Arrays.toString(work.y))
    import skynet.cluster.actors.Profiler.CompletionMessage.CompletionStatus._
    message match {
      case MINIMAL =>
        report(work)
      case EXTENDABLE =>
        split(work)
      case FALSE =>
      // Ignore
      case FAILED =>
        assign(work)
    }
    this.assign(worker)
  }

  private def report(work: Worker.WorkMessage): Unit = {
    this.log.info("UCC: {}", util.Arrays.toString(work.x))
  }

  private def split(work: Worker.WorkMessage): Unit = {
    val x = work.x
    val y = work.y
    val next = x.length + y.length
    if (next < task.attributes - 1) {
      val xNew = util.Arrays.copyOf(x, x.length + 1)
      xNew(x.length) = next
      assign(Worker.WorkMessage(xNew, y))
      val yNew = util.Arrays.copyOf(y, y.length + 1)
      yNew(y.length) = next
      assign(Worker.WorkMessage(x, yNew))
    }
  }
}